contrib/epee/include/net/levin_protocol_handler_async.h

```diff
@@ -190,115 +190,102 @@ class async_protocol_handler
-          epee::span<const uint8_t> fake;
-          cb(LEVIN_ERROR_CONNECTION_TIMEDOUT, fake, con.get_context_ref());
-          con.close();
-          con.finish_outer_call();
-        });
-        m_timer_started = true;
-      }
-    }
-    virtual ~anvoke_handler()
-    {}
-    callback_t m_cb;
-    async_protocol_handler& m_con;
+  protected:
     boost::asio::deadline_timer m_timer;
-    bool m_timer_started;
-    bool m_cancel_timer_called;
-    bool m_timer_cancelled;
-    uint64_t m_timeout;
-    int m_command;
-    virtual bool handle(int res, const epee::span<const uint8_t> buff, typename async_protocol_handler::connection_context& context)
+    async_protocol_handler& m_con;
+    const uint64_t m_timeout;
+    const int m_command;
+
+    virtual void do_handle(int res, const epee::span<const uint8_t> buff)=0;
+
+  public:
+    invoke_response_handler_base(uint64_t timeout, async_protocol_handler& con, int command)
+      : m_timer(con.m_pservice_endpoint->get_io_service()),
+        m_con(con),
+        m_timeout(timeout),
+        m_command(command)
+    {}
+
+    virtual ~invoke_response_handler_base() {}
+
+    //! Run user handler. Only call if `cancel_timer()` returns true.
+    void handle(int res, const epee::span<const uint8_t> buff)
     {
-      if(!cancel_timer())
-        return false;
-      m_cb(res, buff, context);
+      do_handle(res, buff);
       m_con.finish_outer_call();
-      return true;
     }
-    virtual bool is_timer_started() const
+
+    //! Cancel timer and run user handler if not previously invoked \return True if user handler invoked
+    bool cancel()
     {
-      return m_timer_started;
+      if (!cancel_timer())
+        return false;
+      handle(LEVIN_ERROR_CONNECTION_DESTROYED, nullptr);
+      return true;
     }
-    virtual void cancel()
+
+    //! Cancel timer but do not run user handler. \return True if user handler has not been invoked
+    bool cancel_timer()
     {
-      if(cancel_timer())
-      {
-        epee::span<const uint8_t> fake;
-        m_cb(LEVIN_ERROR_CONNECTION_DESTROYED, fake, m_con.get_context_ref());
-        m_con.finish_outer_call();
-      }
+      boost::system::error_code ignored{};
+      return m_timer.cancel(ignored);
     }
-    virtual bool cancel_timer()
+
+    // Attempt to adjust timeout for user handler execution. \return True if timer adjusted.
+    static bool reset_timer(std::shared_ptr<invoke_response_handler_base> self, const bool initial = false)
     {
-      if(!m_cancel_timer_called)
+      if (!self || (!initial && !self->cancel_timer()))
+        return false;
+      if (initial && !self->m_con.start_outer_call())
+        return false;
+
+      auto& self_ = *self;
+      self_.m_timer.expires_from_now(boost::posix_time::milliseconds(self_.m_timeout));
+      self_.m_timer.async_wait([self = std::move(self)](const boost::system::error_code& ec)
```

Comment:
For example, if `async_wait()` were to move the lambda into another function, then after that function call tried to access a data member of `self_.m_timer`, the data member won't exist because the lambda would be destroyed in the function call, so there will be a segfault.

Reply:
Crap, I can't believe I missed this. The safety of this is dependent on how `async_wait` is implemented internally (technically the only callee ensures safety here, but this is easy to miss in a refactor).

I did this elsewhere, and it definitely needs to be corrected in a couple of places (incoming PR!). Some of the existing cases uses an `asio::strand` though, and so you can "get away with it" if the `std::move` is done inside the strand (because the next op has to wait in the queue until the current instance returns).

Reply:
After further thought, I think this is valid, but I'm still likely to change it.

There are plenty of examples of a single async callback starting another instance of itself in a "loop". The documentation states that no synchronization is needed because it is an implicit single strand of calls. If this line here is an issue, then those types of setups would also be an issue.

I looked at some of the implementations for the async functions, and they all seem to never touch the internals after queuing an operation. It's probably for the "implicit" strand case; I could find no examples of this exact setup.

Reply:
You could pass `self` as `std::weak_ptr`: https://floating.io/2017/07/lambda-shared_ptr-memory-leak/

Reply:
It might be worth changing (despite the small bump in CPU cycles), because its a better guarantee that no memory leak occurs. The callback is invoked so infrequently, that I will just go ahead and make that change too.

I don't see how `release_protocol()` can be removed though, otherwise I would've make this a `std::weak_ptr` already. The current behavior is to invoke `handle(...)` when destroying the connection, and doing this automatically in destructor cleanup is iffy because `m_connection_context` could already be destroyed.